src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -285,16 +263,128 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
                 context.Write(writer, new DynamicJsonValue
                 {
                     [nameof(BatchCommandResult.Results)] = array,
-                    [nameof(BatchCommandResult.TransactionIndex)] = result.Index
+                    [nameof(BatchCommandResult.TransactionIndex)] = index
                 });
             }
         }
 
+        private async Task<long?> GetClusterTransactionCount(object result, string raftRequestId, long databaseCommandsCount, Task<long?> onDatabaseCompletionTask)
+        {
+            if (databaseCommandsCount == 0)
+                return null;
+
+            long? count = null;
+
+            if (result is List<ClusterTransactionCommand.ClusterTransactionErrorInfo> errors)
+                ThrowClusterTransactionConcurrencyException(errors);
+
+            if (result != null && result is ClusterTransactionResult clusterTxResult)
+            {
+                if(clusterTxResult.Count.HasValue == false)
+                    throw new InvalidOperationException($"Cluster Transaction result is null, but has more then 0 database commands.");
+
+                count = clusterTxResult.Count.Value;
+            
+                Database.ForTestingPurposes?.AfterCommitInClusterTransaction?.Invoke();
+            
+                using (var token = CreateHttpRequestBoundOperationToken())
+                    await onDatabaseCompletionTask.WithCancellation(token.Token);
+            
+            }
+            else // leader isn't updated
+            {
+                using (var token = CreateHttpRequestBoundOperationToken())
+                    count = await onDatabaseCompletionTask.WithCancellation(token.Token);
+
+                if (count.HasValue == false) // Failover and the ClusterTransactionCommand is already completed in the DocumentDatabase, so the task with the result is no longer exists, so we'll try to take the result from the local history log.
+                {
+                    using (ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
+                    using (ctx.OpenReadTransaction())
+                    {
+                        // searching in the history log
+                        if (ServerStore.Engine.LogHistory.TryGetCommandByGuid(ctx, raftRequestId, out var localCommand)
+                            && localCommand.TryGet(nameof(LogHistoryColumn.Result), out BlittableJsonReaderObject ctlrBlt))
```

Comment:
If you find in the history log but the result doesn't have `LogHistoryColumn.Result` in it you throw the exception in the `else`.
That can be confusing, no?
If I understand it right the odds for that are low but in this case maybe better to use the index operator:
`var ctlrBlt = localCommand[nameof(LogHistoryColumn.Result)];`
So in case it doesn't have this property we will get the right exception

Reply:
That cannot happen because you are getting here after the command is committed and applied, and completed in the db, so in the worst case you'll never find it in the log because it has been already deleted from there..
But if you **found** the command it has to be completed and have a result (because we waiting for the command completion before we get here).
So you cant get here and really get 'false' from the 'TryGet'..

Reply:
BTW, the exception we throw is fit to that case also (the case when the 'TryGet' will return 'false'), but the case is impossible.

Reply:
I added another exception for that case as well..

Reply:
How does the exception fit when `TryGet` returns false?
You found the entry in the table but the entry does not contain the `Result` property, no?
I do understand that this situation can't happen.
Unless you have a bug then we will get the wrong error.
**Why did you add this check (`TryGet`) if it can't happen?** 
You can put the result of the check in `Debug.Assert` if you don't expect it to happen.
But, if I understand it right, adding it to the `TryGetCommandByGuid` check is not relevant.

Reply:
I prefer that it will throw that expression (the message is a little bit different.. look at the end of it) than the indexer will throw.